Reth Book - Stages framework/draft (#360)
paradigmxyz/reth: book/stages/README.md

# Stages

The `stages` lib plays a central role in syncing the node, maintaining state, updating the database and more. The stages involved in the Reth pipeline are the `HeaderStage`, `BodyStage`, `SendersStage`, and `ExecutionStage` (note that this list is non-exhaustive, and more pipeline stages will be added in the near future). Each of these stages are queued up and stored within the Reth pipeline.

[Filename: crates/stages/src/pipeline.rs](https://github.com/paradigmxyz/reth/blob/main/crates/stages/src/pipeline.rs#L76)
```rust
pub struct Pipeline<DB: Database> {
    stages: Vec<QueuedStage<DB>>,
    max_block: Option<BlockNumber>,
    events_sender: MaybeSender<PipelineEvent>,
}
```

When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 

> Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution. 

Reviewer: ```suggestion
Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the stage if there was an issue during the stage execution. 
```
nit: do you mean to say "state" here? not all stages touch EVM "state" so this may be a slightly confusing term

Author: What I meant by this is `unwind the changes to the database`. I can update the language to reflect this or reword it to whatever makes the most sense.

[Filename: crates/stages/src/stage.rs](https://github.com/paradigmxyz/reth/blob/main/crates/stages/src/stage.rs#L64)
```rust
#[async_trait]
pub trait Stage<DB: Database>: Send + Sync {
    /// Get the ID of the stage.
    ///
    /// Stage IDs must be unique.
    fn id(&self) -> StageId;

    /// Execute the stage.
    async fn execute(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: ExecInput,
    ) -> Result<ExecOutput, StageError>;

    /// Unwind the stage.
    async fn unwind(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: UnwindInput,
    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
}
```

To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.

<br>

## HeadersStage

> The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.

Reviewer: This might be a good place to talk about why the headers stage fetches headers in descending order (as opposed to ascending order, like the other stages), to avoid long-range attacks

Each value yielded from the stream is a `SealedHeader`. 

[File: crates/primitives/src/header.rs](https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/header.rs#L207)
```rust
/// A [`Header`] that is sealed at a precalculated hash, use [`SealedHeader::unseal()`] if you want
/// to modify header.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SealedHeader {
    /// Locked Header fields.
    header: Header,
    /// Locked Header hash.
    hash: BlockHash,
}
```

> Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and a full header validation adhering to the consensus spec occurs in the `downloader`. Following the basic header validation, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the db changes are unwound and the stage is resumed from the most recent valid state.

Reviewer: validation according to consensus is going to stick out to the reader, and at a glance this might be read as "this stage does not do validation according to consensus". additionally, it's not super clear what the `downloader` is here.

i think it'd be helpful to mention, either here or in the paragraph above, that the stream that is established is an implementation of the `HeaderDownloader` trait (and, I suppose, pull in the definition for that trait), and double-down on the point that the implementation of that trait must validate the downloaded header according to the consensus specification.

so basically we construct a `SealedHeader` for the current latest header, then we request from our peers to download (in reverse order) all headers since the `SealedHeader`, and it is in the implementation of the downloader that consensus validation occurs

Reviewer: sorry, i know that was a super verbose comment, just wanted to get it down - please let me know if any part of it could be clarified

This process continues until all of the headers have been downloaded and and written to the database. Finally, the total difficulty of the chain's head is updated and the function returns `Ok(ExecOutput { stage_progress: current_progress, reached_tip: true, done: true })`, signaling that the header sync has completed successfully. 

<br>

## BodyStage

Once the `HeadersStage` completes successfully, the `BodyStage` will start execution. The body stage downloads block bodies for all of the new block headers that were stored locally in the database. The `BodyStage` first determines which block bodies to download by checking if the block body has an ommers hash and transaction root. 

> An ommers hash is the Keccak 256-bit hash of the ommers list portion of the block. If you are unfamiliar with ommers blocks, you can [click here to learn more](https://ethereum.org/en/glossary/#ommer). Note that while ommers blocks were important for new blocks created during Ethereum's proof of work chain, Ethereum's proof of stake chain selects exactly one block proposer at a time, causing ommers blocks not to be needed in post-merge Ethereum.

Reviewer: this is a great explanation to add, nice

[File: crates/primitives/src/transaction/signature.rs](https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/transaction/signature.rs#L72)
```rust

    /// Recover signature from hash.
    pub(crate) fn recover_signer(&self, hash: H256) -> Option<Address> {
        let mut sig: [u8; 65] = [0; 65];

        self.r.to_big_endian(&mut sig[0..32]);
        self.s.to_big_endian(&mut sig[32..64]);
        sig[64] = self.odd_y_parity as u8;

        secp256k1::recover(&sig, hash.as_fixed_bytes()).ok()
    }
```

In an [ECDSA (Elliptic Curve Digital Signature Algorithm) signature](), the "r", "s", and "v" values are three pieces of data that are used to mathematically verify the authenticity of a digital signature. ECDSA is a widely used algorithm for generating and verifying digital signatures, and it is often used in cryptocurrencies like Ethereum.

> The "r" is the x-coordinate of a point on the elliptic curve that is calculated as part of the signature process. The "s" is the s-value that is calculated during the signature process. It is derived from the private key and the message being signed. Lastly, the "v" is the "recovery value" that is used to recover the public key from the signature, which is derived from the signature and the message that was signed. Together, the "r", "s", and "v" values make up an ECDSA signature, and they are used to verify the authenticity of the signed transaction.

Reviewer: Is there a resource we could link to here that goes more deeply into depth on ECDSA signatures, if readers are interested? (doesn't need to be in this PR)

Reviewer: maybe in a further version of the docs we can allude to EIP-155 changes here? idk, maybe too low level, don't have much preference

[Filename: crates/stages/src/stages/execution.rs](https://github.com/paradigmxyz/reth/blob/main/crates/stages/src/stages/execution.rs#L222)
```rust
block_change_patches.push((
    reth_executor::executor::execute_and_verify_receipt(
        header,
        &recovered_transactions,
        &self.config,
        &mut state_provider,
    )
    .map_err(|error| StageError::ExecutionError { block: header.number, error })?,
    start_tx_index,
    block_reward_index,
));
```

After all headers and their corresponding transactions have been executed, all of the resulting state changes are applied to the database, updating account balances, account bytecode and other state changes. After applying all of the execution state changes, if there was a block reward, it is applied to the validator's account. 

At the end of the `execute()` function, a familiar value is returned, `Ok(ExecOutput { done: is_done, reached_tip: true, stage_progress: last_block })` signaling a successful completion of the `ExecutionStage`.

<br>

# Next Chapter

> Now that we have covered all of the stages that are currently included in the `Pipeline`, you know how the Reth client stays synced with the chain tip and updates the database with all of the new headers, bodies, senders and state changes. While this chapter provides an overview, on how the pipeline stages work, the following chapter will dive deeper into the database, the networking stack and other exciting corners of the Reth codebase. Feel free to check out any of the codebase mentioned in this chapter and when you are ready, the next chapter will dive into the `database`.

Reviewer: ```suggestion
Now that we have covered all of the stages that are currently included in the `Pipeline`, you know how the Reth client stays synced with the chain tip and updates the database with all of the new headers, bodies, senders and state changes. While this chapter provides an overview on how the pipeline stages work, the following chapters will dive deeper into the database, the networking stack and other exciting corners of the Reth codebase. Feel free to check out any parts of the codebase mentioned in this chapter, and when you are ready, the next chapter will dive into the `database`.
```
(small nits)

/// while a [HeadersClient] represents a client capable of fulfilling these requests.
#[auto_impl::auto_impl(&, Arc, Box)]
pub trait HeaderDownloader: Sync + Send + Unpin {
    /// The Consensus used to verify block validity when
    /// downloading
    type Consensus: Consensus;

    /// The Client used to download the headers
    type Client: HeadersClient;

    /// The request timeout duration
    fn timeout(&self) -> Duration;

    /// The consensus engine
    fn consensus(&self) -> &Self::Consensus;

    /// The headers client
    fn client(&self) -> &Self::Client;

    /// Download the headers
    fn download(&self, head: SealedHeader, forkchoice: ForkchoiceState) -> HeaderBatchDownload<'_>;

    /// Stream the headers
    fn stream(&self, head: SealedHeader, forkchoice: ForkchoiceState) -> HeaderDownloadStream;

    /// Validate whether the header is valid in relation to it's parent
    ///
    /// Returns Ok(false) if the
    fn validate(&self, header: &SealedHeader, parent: &SealedHeader) -> Result<(), DownloadError> {
        validate_header_download(self.consensus(), header, parent)?;
        Ok(())
    }
}
```

> The `HeaderStage` relies on the downloader stream to return the headers in descending order staring from the chain tip down to the latest block in the database. While other stages in the `Pipeline` start from the most recent block in the database up to the chain tip, the `HeaderStage` works in reverse to avoid [long-range attacks](https://messari.io/report/long-range-attack). When a node downloads headers in ascending order, it will not know if it is being subjected to a long-range attack until it reaches the most recent blocks. To combat this, the `HeaderStage` starts by getting the chain tip from the Consensus Layer, verifies the tip, and then walks backwards by the parent hash. Each value yielded from the stream is a `SealedHeader`. 

Reviewer: ```suggestion
The `HeaderStage` relies on the downloader stream to return the headers in descending order starting from the chain tip down to the latest block in the database. While other stages in the `Pipeline` start from the most recent block in the database up to the chain tip, the `HeaderStage` works in reverse to avoid [long-range attacks](https://messari.io/report/long-range-attack). When a node downloads headers in ascending order, it will not know if it is being subjected to a long-range attack until it reaches the most recent blocks. To combat this, the `HeaderStage` starts by getting the chain tip from the Consensus Layer, verifies the tip, and then walks backwards by the parent hash. Each value yielded from the stream is a `SealedHeader`. 
```
nit: typo

[File: crates/primitives/src/header.rs](https://github.com/paradigmxyz/reth/blob/main/crates/primitives/src/header.rs#L207)
```rust
/// A [`Header`] that is sealed at a precalculated hash, use [`SealedHeader::unseal()`] if you want
/// to modify header.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SealedHeader {
    /// Locked Header fields.
    header: Header,
    /// Locked Header hash.
    hash: BlockHash,
}
```

> Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and the `HeadderDownloader` uses the `validate` method during the `stream`, so that each header is validated according to the consensus specification before the header is yielded from the stream. After this, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the changes to the database are unwound and the stage is resumed from the most recent valid state.

Reviewer: ```suggestion
Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and the `HeaderDownloader` uses the `validate` method during the `stream`, so that each header is validated according to the consensus specification before the header is yielded from the stream. After this, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the changes to the database are unwound and the stage is resumed from the most recent valid state.
```
nit: typo

# Stages

> The `stages` lib plays a central role in syncing the node, maintaining state, updating the database and more. The stages involved in the Reth pipeline are the `HeaderStage`, `BodyStage`, `SendersStage`, and `ExecutionStage`. Each of these stages are queued up and stored within the Reth pipeline.

Reviewer 2: pls add a comment that this list is nonexhaustive, more will be added shortly

Author: Sounds good, just added a note about this.

Filename: crates/stages/src/pipeline.rs
```rust
pub struct Pipeline<DB: Database> {
    stages: Vec<QueuedStage<DB>>,
    max_block: Option<BlockNumber>,
    events_sender: MaybeSender<PipelineEvent>,
}
```

> When the node is first started, a new `Pipeline` is intialized and all of the stages are added into `Pipline.stages`. Then, the `Pipeline::run` function is called, which starts the pipline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 

Reviewer 2: ```suggestion
When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 
```

Author: Thanks for catching these, I added a spellchecker to my IDE so I can catch these earlier in the future.

Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution.

Filename: crates/stages/src/stage.rs
```rust
#[async_trait]
pub trait Stage<DB: Database>: Send + Sync {
    /// Get the ID of the stage.
    ///
    /// Stage IDs must be unique.
    fn id(&self) -> StageId;

    /// Execute the stage.
    async fn execute(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: ExecInput,
    ) -> Result<ExecOutput, StageError>;

    /// Unwind the stage.
    async fn unwind(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: UnwindInput,
    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
}
```

> To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execution()` function at each stage, starting with `HeadersStage`.

Reviewer 2: ```suggestion
To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.
```

## HeadersStage

The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.

> Each value produced from the stream is a `Vec<SealedHeader>`. 

Reviewer 2: stream yields `SealedHeader` (not vec)

# Stages

The `stages` lib plays a central role in syncing the node, maintaining state, updating the database and more. The stages involved in the Reth pipeline are the `HeaderStage`, `BodyStage`, `SendersStage`, and `ExecutionStage` (note that this list is non-exhaustive, and more pipeline stages will be added in the near future). Each of these stages are queued up and stored within the Reth pipeline.

> Filename: crates/stages/src/pipeline.rs

Reviewer 3: Let's link to the actual file and line number?

Author: Sounds good, I will update this.

```rust
pub struct Pipeline<DB: Database> {
    stages: Vec<QueuedStage<DB>>,
    max_block: Option<BlockNumber>,
    events_sender: MaybeSender<PipelineEvent>,
}
```

When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 

Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution. 

Filename: crates/stages/src/stage.rs
```rust

> #[async_trait]

Reviewer: ooh, i can look into how one might do this, ya

> #[async_trait]

Reviewer: this should do the trick, ya? https://rust-lang.github.io/mdBook/format/mdbook.html#including-portions-of-a-file (i'm thinking we can just add `ANCHOR`s to the source code)

pub trait Stage<DB: Database>: Send + Sync {
    /// Get the ID of the stage.
    ///
    /// Stage IDs must be unique.
    fn id(&self) -> StageId;

    /// Execute the stage.
    async fn execute(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: ExecInput,
    ) -> Result<ExecOutput, StageError>;

    /// Unwind the stage.
    async fn unwind(
        &mut self,
        db: &mut StageDB<'_, DB>,
        input: UnwindInput,
    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
}
```

To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.

## HeadersStage

> The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.

Reviewer 3: We should eventually add cross-links to everything once we have them, e.g. `eth/65 handshakes` shoudl link to the book section that talks about it
